Effects: Add backend for Audio Unit (AU) plugins on macOS

fwcd opened the pull request:
This implements an effects backend for AU plugins on macOS. In other words, all of the stock AU plugins, along with any installed third-party Logic/GarageBand plugins can now conveniently be used directly from Mixxx's effects panel.

<img width="1180" alt="Screenshot 2023-10-15 at 03 39 06" src="https://github.com/mixxxdj/mixxx/assets/30873659/f6db6330-eb21-4dd4-a9c3-96c02861e524">
<img width="1392" alt="Screenshot 2023-10-15 at 03 39 44" src="https://github.com/mixxxdj/mixxx/assets/30873659/733b80e0-69e5-40e5-b645-31cfc36a7a0e">

## Implementation Notes

This leverages the [AUv2 C API](https://developer.apple.com/documentation/audiotoolbox/audio_unit_v3_plug-ins/hosting_audio_unit_extensions_using_the_auv2_api?language=objc) for hosting Audio Units in a standalone fashion, without an `AUGraph`, `AVAudioEngine` or similar. The audio units are instantiated out-of-process (though this can easily be changed) and are passed audio data directly via the render callback mechanism ([documented in detail here](https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/AudioUnitProgrammingGuide/AudioUnitDevelopmentFundamentals/AudioUnitDevelopmentFundamentals.html#//apple_ref/doc/uid/TP40003278-CH7-SW14)).

The rationale for not using the [AUv3 Objective-C API](https://developer.apple.com/documentation/audiotoolbox/audio_unit_v3_plug-ins/migrating_your_audio_unit_host_to_the_auv3_api?language=objc) is that the latter seems to be a bit more finicky to deal with in a standalone context (without `AUGraph`/`AVAudioEngine`), e.g. throwing `NoConnection` errors during rendering. Since the API is supposed to map 1-to-1 to AUv2, it should probably still be possible to express our use case with the AUv3 API, but since the AUv2 API is not deprecated (and also still widely used), we'll use that for now. Note that AUv3 plugins will work fine too, since the AUv2 and AUv3 APIs are bridged in both directions.

The (compiling, but non-functional) implementation using the AUv3 API can be found here:

- https://github.com/fwcd/mixxx/pull/5

## Link dump

Further reading and anything that may be of use to implementors:

- General:
  - https://developer.apple.com/library/archive/documentation/MusicAudio/Conceptual/AudioUnitHostingGuide_iOS/AudioUnitHostingFundamentals/AudioUnitHostingFundamentals.html
- AUv2:
  - https://developer.apple.com/documentation/audiotoolbox/audio_unit_v2_c_api?language=objc
  - https://developer.apple.com/documentation/audiotoolbox/audio_unit_v3_plug-ins/hosting_audio_unit_extensions_using_the_auv2_api?language=objc
  - https://stackoverflow.com/questions/56778707/does-augraph-deprecation-means-no-more-audio-render-callbacks
  - https://stackoverflow.com/questions/6735444/ios-audio-units-when-is-usage-of-augraphs-necessary
- AUv3:
  - https://developer.apple.com/documentation/audiotoolbox/audio_unit_v3_plug-ins
  - https://developer.apple.com/documentation/audiotoolbox/audio_unit_v3_plug-ins/migrating_your_audio_unit_host_to_the_auv3_api?language=objc
  - https://developer.apple.com/forums/thread/14458
  - https://github.com/hotpaw2/auv3test5

## Prior art

- Implementation of Audio Unit effects in Audacity (also uses the AUv2 API): https://github.com/audacity/audacity/blob/52708c770ffcd98cb86527cd7756413c4179f7fe/libraries/lib-audio-unit/AudioUnitInstance.cpp

## Caveats

The plugins currently have no UI. Might be a nice candidate for a future PR, but for now they have to be controlled via the parameter knobs.

fwcd committed:
AUBackend: Stub out Audio Unit (AU)-based effects backend

fwcd committed:
AUBackend: Move class into source file

Making the class opaque lets us store Objective-C objects there.

fwcd committed:
AUBackend: Move implementations into class

fwcd committed:
AUBackend: Stub out effect processor and manifest

fwcd committed:
AUManifest: Store AVAudioUnitComponent

fwcd committed:
AUBackend: Find available audio units

fwcd committed:
AUBackend: Log found audio units

fwcd committed:
EffectsBackend: Add backend type name for AU (Audio Unit)

fwcd committed:
AUManifest: Set correct backend type

fwcd committed:
AUManifest: Pass component to constructor

fwcd committed:
AUManifest: Expose more metadata about the components

fwcd committed:
AUBackend: Store manifests

fwcd committed:
EffectsBackend: Remove space from 'Audio Unit'

Turns out, the parser unfortunately cannot really deal with this,
`EffectsBackendManager::getManifestFromUniqueId` will then only split
off a a part of the backend type name (in the case auf 'Audio Unit' just
'Unit'), which would then incorrectly parse to the built-in type and
subsequently not find the manifest.

fwcd committed:
AUBackend: Identify AU effects by their metadata

fwcd committed:
AUEffectProcessor: Implement basic audio unit instantiation

fwcd committed:
AUEffectProcessor: Move instantiation to AudioUnitManager

fwcd committed:
AUEffectProcessor: Try implementing the AU effect render cycle

fwcd committed:
AUEffectProcessor: Perform (de)allocation only once

fwcd committed:
AUEffectProcessor: Check render status

fwcd committed:
AUEffectProcessor: Add more nullability annotations

fwcd committed:
AUEffectProcessor: Configure format

fwcd committed:
AUEffectProcessor: Configure audio unit busses

fwcd committed:
AUEffectProcessor: Factor out AudioUnitManager

fwcd committed:
AUEffectProcessor: Use AUv2 API

fwcd committed:
AUEffectProcessor: Pass the actual sample count

fwcd committed:
AudioUnitManager: Dispose of audio unit properly

fwcd committed:
AudioUnitManager: Add option to instantiate synchronously

fwcd committed:
AudioUnitManager: Enable sync instantiation

fwcd committed:
AudioUnit: Rename AU -> AudioUnit

fwcd committed:
AudioUnitManager: Update description

fwcd committed:
AudioUnit: Use long form for everything

Both to avoid clashes with Apple's AU-prefixed classes and to improve
clarity, especially for contributors that might not be familiar with
this acronym.

fwcd committed:
AudioUnitManifest: Discover parameters

fwcd committed:
AudioUnitManifest: Add parameters

fwcd committed:
AudioUnitManifest: Apply scaling to parameters

fwcd committed:
AudioUnitEffectProcessor: Try implementing parameter loading

fwcd committed:
AudioUnitEffectProcessor: Implement parameter syncing

fwcd committed:
AudioUnitEffectProcessor: Add missing nullability annotations

fwcd committed:
AudioUnitBackend: Remove redundant includes

deejay-girl commented:
I don't have a Mac but as I understand correctly on Windows those are called VSTs?
Would it be possible to have the same feature for VSTs?
I found online a Mac can run VSTs too.

fwcd commented:
Well, yes and no, AUs and VSTs are similar in that they are both audio plugins, but they're completely different implementations, so no, they aren't interchangeable.

In principle the effects framework provides a general interface that could be implemented for VSTs too ([see the wiki](https://github.com/mixxxdj/mixxx/wiki/Effects-Framework)). Since AUs use a completely different API, support for VSTs would have to be implemented "from scratch" (which is a fair bit of work too and definitely out-of-scope for this PR).

fwcd committed:
AudioUnitEffectProcessor: Remove redundant field

fwcd committed:
AudioUnitManifest: Link the first parameter by default

fwcd committed:
AudioUnitEffectProcessor: Set stream format explicitly

fwcd committed:
AudioUnitEffectProcessor: Set stream format on input and output scope

m0dB commented:
I will try to review this over the weekend.

m0dB commented on src/effects/backends/audiounit/audiounitmanager.mm:
```diff
@@ -0,0 +1,109 @@
+    // We need to load this atomic flag to ensure that we don't get a partial
+    // read of the audio unit pointer (probably extremely uncommon, but not
+    // impossible: https://belkadan.com/blog/2023/10/Implicity-Atomic)
+    if (!m_isInstantiated.load()) {
```
Are you using an atomic instead of a mutex because this method is called from the realtime audio thread? If so, a comment mentioning that would be good.

fwcd replied:
That was what I assumed, `getAudioUnit` is called from `processChannel` which operates directly on sample buffers, though it might be worth checking whether that is indeed the audio thread. Since that's a method on the hot path, we want to avoid blocking there.

fwcd replied:
Added a short note in 440d1cf4b4.

m0dB commented:
This is really nice! The code looks good to me, but I would like to do a bit of testing before approving.

m0dB commented:
Unfortunately it isn't working here. I tried a bunch of the Apple default AudioUnits. The logs are outputting a stream of `warning [Engine] Rendering Audio Unit failed with status -10874` messages, and the audio output is stuck on a 1024 sample buffer.

Note that this is after rebasing, compilation of the branch as is failed in unrelated code.

fwcd commented:
I'll have a look, it's been a while since I touched this PR, might even have been before the Sonoma update

fwcd commented:
https://osstatus.com/search/results?platform=all&framework=all&search=-10874 says this is `kAudioUnitErr_TooManyFramesToProcess`. Maybe it needs a smaller buffer

fwcd commented:
The branch is now rebased onto the most recent `main`, so please reset any local clone. I can confirm that the Audio Unit effects currently only work if the audio buffer is <= 10.7 ms. We probably have to perform multiple rendering passes over fixed-sized chunks if the buffer is larger.

fwcd committed:
AudioUnitEffectProcessor: Process buffer in chunks

This fixes an issue where rendering the audio unit would fail if the
audio buffer was > 10.7 ms.

fwcd committed:
AudioUnitManager: Add rationale for not using a mutex

JoergAtGithub commented:
@m0dB Could you test this again?

m0dB commented:
I confirm it works fine with 11.6 ms @ 44,100 Hz .

I don't think it's a good idea to merge this before taking care of handling larger buffer sizes with multiple render passes. Did you look into implementing that, @fwcd? Or would you like me to give it a try?

fwcd commented:
IIRC I did implement that in https://github.com/mixxxdj/mixxx/pull/12112/commits/f9cfd6295c86a4fcd87c3b87e5ea75b8d587ef3c 😄

m0dB commented:
Will you add that to this PR then?

JoergAtGithub commented:
> Will you add that to this PR then?

This commit is already part of this PR!

m0dB commented:
Looks like I somehow tested without that commit last time. I retested and it indeed works fine!

JoergAtGithub commented:
Code LGTM! Thank you!

fwcd commented:
Can/should we add this to the change log?

JoergAtGithub commented:
The label changelog is set for this. When the next changelog PR will be opened, at leat all PRs with changelog label should be added to the changelog. Feel free to take over this task.
